tokio/src/sync/notify.rs

```diff
@@ -0,0 +1,496 @@
+
+                    // Transition the state to WAITING.
+                    loop {
+                        match curr {
+                            EMPTY => {
+                                // Transition to WAITING
+                                let res = notify
+                                    .state
+                                    .compare_exchange(EMPTY, WAITING, SeqCst, SeqCst);
+
+                                if let Err(actual) = res {
+                                    curr = actual;
+                                    continue;
+                                }
+                            }
+                            WAITING => {}
+                            NOTIFIED => {
+                                // Try consuming the notification
+                                let res = notify
+                                    .state
+                                    .compare_exchange(NOTIFIED, EMPTY, SeqCst, SeqCst);
+
+                                match res {
+                                    Ok(_) => {
+                                        // Acquired the notification
+                                        *state = Done;
+                                        continue 'outer;
+                                    }
+                                    Err(actual) => {
+                                        curr = actual;
+                                        continue;
+                                    }
+                                }
+                            }
+                            _ => unreachable!(),
+                        }
+
+                        // The state has been transitioned to waiting
+                        break;
+                    }
+
+                    // Safety: called while locked.
+                    unsafe {
+                        (*waiter.as_mut().get()).waker = Some(cx.waker().clone());
+
+                        // Insert the waiter into the linked list
+                        waiters.push_front(waiter.as_mut());
+                    }
+
+                    *state = Waiting;
+                }
+                Waiting => {
+                    let mut waiters = notify.waiters.lock().unwrap();
+
+                    {
+                        // Safety: called while locked
+                        let w = unsafe { &mut *waiter.as_mut().get() };
+
+                        if w.notified {
+                            w.waker = None;
+                            w.notified = false;
+
+                            // Remove the entry from the linked list. The entry
+                            // may only be in the linked list while the state is
+                            // `Waiting`.
+                            //
+                            // Safety: called while the lock is held
+                            unsafe {
+                                waiters.remove(waiter.as_mut());
```

Comment:
I'm confused -- isn't the `waiter` removed by the call to `pop_back` in the notify code? Why does it _also_ need to be removed here?